File: android/app/src/main/java/com/vaultsync/domain/usecase/GetSyncStatusUseCase.kt
package com.vaultsync.domain.usecase

import com.vaultsync.domain.model.SyncStatus
import com.vaultsync.domain.model.SyncStatusSummary
import com.vaultsync.domain.repository.DocumentRepository
import com.vaultsync.domain.repository.SyncController
import com.vaultsync.domain.repository.SyncRepository
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.combine

class GetSyncStatusUseCase(
    private val documentRepository: DocumentRepository,
    private val syncRepository: SyncRepository,
    private val syncController: SyncController
) {
    operator fun invoke(): Flow<SyncStatusSummary> {
        return combine(
            documentRepository.observeDocuments(),
            syncRepository.observeConflicts(),
            syncController.isSyncing
        ) { docs, conflicts, syncing ->
            val synced = docs.count { it.syncStatus == SyncStatus.SYNCED }
            val pending = docs.count { it.syncStatus == SyncStatus.PENDING || it.syncStatus == SyncStatus.SYNCING }
            SyncStatusSummary(
                totalDocuments = docs.size,
                syncedCount = synced,
                pendingCount = pending,
                conflictCount = conflicts.size,
                isSyncing = syncing
            )
        }
    }

    suspend fun executeDirect(): SyncStatusSummary {
        val docs = documentRepository.getDocuments()
        val conflicts = syncRepository.getConflicts()
        val synced = docs.count { it.syncStatus == SyncStatus.SYNCED }
        val pending = docs.count { it.syncStatus == SyncStatus.PENDING || it.syncStatus == SyncStatus.SYNCING }
        return SyncStatusSummary(
            totalDocuments = docs.size,
            syncedCount = synced,
            pendingCount = pending,
            conflictCount = conflicts.size,
            isSyncing = syncController.isSyncing.value
        )
    }
}

File: ios/Sources/VaultSync/Domain/UseCases/GetSyncStatusUseCase.swift
import Foundation

public final class GetSyncStatusUseCase: Sendable {
    private let documentRepository: any DocumentRepository
    private let syncRepository: any SyncRepository
    private let syncController: any SyncController

    public init(
        documentRepository: any DocumentRepository,
        syncRepository: any SyncRepository,
        syncController: any SyncController
    ) {
        self.documentRepository = documentRepository
        self.syncRepository = syncRepository
        self.syncController = syncController
    }

    public func executeDirect() async throws -> SyncStatusSummary {
        let docs = try await documentRepository.getDocuments()
        let conflicts = try await syncRepository.getConflicts()

        let synced = docs.filter { $0.syncStatus == .synced }.count
        let pending = docs.filter { $0.syncStatus == .pending || $0.syncStatus == .syncing }.count

        return SyncStatusSummary(
            totalDocuments: docs.count,
            syncedCount: synced,
            pendingCount: pending,
            conflictCount: conflicts.count,
            isSyncing: false
        )
    }
}
